### PR TITLE
4 user registration page

### PR DESCRIPTION
**Pull request by Brianisthebest**
# Description
- Added form on register page to create new user
- Added validations to user model for presence of name and email and uniqueness of email
- Happy and sad path testing for if email has been used or not

## Type of change

- [ ] Bug fix
- [x] New feature
- [ ] Database Change
- [ ] Refactor

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes.

- [x] Testing to see if form is visible
- [x] Testing to fill form in with already used email to redirect with flash message
- [x] Testing to fill in form with usable email with redirect to user show page
- [x] Updated test for landing page to account for link instead of button. BUTTONS DON'T WANT TO WORK WITH ME

# Checklist:

- [x] Tests created for new features pass
- [x] All tests pass (This was a good one to throw on the checklist, Grace, I would have missed the fact I changed my button to a link.

### CODE REVIEW

File: app/views/users/new.html.erb [hunk 1]
```diff
@@ -0,0 +1,9 @@
+<h2>Register a New User</h2>
+
+<%= form_with url: register_path, method: :post, local: true do |form| %>
+  <%= form.label :user_name %>
+  <%= form.text_field :user_name %><br>
+  <%= form.label :email %>
+  <%= form.text_field :email %><br>
+  <%= form.submit "Create New User" %>
+<% end %>
```
Comment (grace-joh) [inline app/views/users/new.html.erb:9]: just a thought to refactor in the future with a model binded(?) bound(?) to this form!

File: spec/features/users/new_spec.rb [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+      expect(page).to have_field('Email')
+      expect(page).to have_button('Create New User')
+    end
+
+    it 'when I fill out the email with one that already exists I should see an error message' do
+      create(:user, email: 'already_here@hotmail.edu')
+
+      fill_in 'User name', with: 'Test User'
+      fill_in 'Email', with: 'already_here@hotmail.edu'
+      click_button 'Create New User'
```
Comment (grace-joh) [inline spec/features/users/new_spec.rb:21]: just noting the `click button` here

File: spec/features/users/new_spec.rb [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+      expect(page).to have_content('Email has already been taken')
+    end
+
+    it 'when I fill out the email with a new one I be redirected to the user show page for that user' do
+
+      fill_in 'User name', with: 'Test User'
+      fill_in 'Email', with: 'new_email@hotmail.edu'
+      click_button 'Create New User'
```
Comment (grace-joh) [inline spec/features/users/new_spec.rb:30]: just noting the `click button` here

**Review by grace-joh — CHANGES_REQUESTED** [review]
For the Landing page, it looks like you started with a button and then changed to a link! I think you forgot to update the test?

File: app/views/landing/index.html.erb [hunk 4]
```diff
@@ -1,5 +1,5 @@
 <div id="new-user">
-  <%= button_to "Create User", register_path %>
+  <%= link_to "Create User", register_path, method: :get, local: true %>
```
[the remark this answers is not in the document]
Comment (Brianisthebest) [reply app/views/landing/index.html.erb:2]: The button is for the form that creates the user and the link is to go to the form. I couldn't get the button to register the get method so I just swapped it to a link.

[the remark this answers is not in the document]
Comment (grace-joh) [reply app/views/landing/index.html.erb:2]: Ah i see! Good call!

File: spec/features/users/new_spec.rb [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+  before(:each) do
+    visit '/register'
+  end
+
+  describe 'when I visit the registration page' do
+    it 'I should see a form with a text field for name and email and a button to register' do
+      expect(page).to have_content('Register a New User')
+      expect(page).to have_field('User name')
+      expect(page).to have_field('Email')
+      expect(page).to have_button('Create New User')
```
[the remark this answers is not in the document]
Comment (grace-joh) [reply spec/features/users/new_spec.rb:13]: nope this was a button oops my b!

**Pull request merged**